src/syscall_mocking/mod.rs

```diff
@@ -45,33 +47,73 @@ impl Tracer {
         }
     }
 
-    fn execve(executable: &Path, args: Vec<String>, env: HashMap<String, String>) -> R<()> {
-        let path = CString::new(executable.as_os_str().as_bytes())?;
-        let mut c_args = vec![path.clone()];
-        for arg in args {
-            c_args.push(CString::new(arg)?);
+    fn execve_params(
+        interpreter: Option<Vec<u8>>,
+        program: &Path,
+        args: Vec<String>,
+        env: HashMap<String, String>,
+    ) -> R<(CString, Vec<CString>, Vec<CString>)> {
+        let c_executable = CString::new(program.as_os_str().as_bytes())?;
+        let mut c_args = VecDeque::new();
+        c_args.push_back(c_executable.clone());
+        for arg in &args {
+            c_args.push_back(CString::new(arg.clone())?);
         }
         let mut c_env = vec![];
         for (key, value) in env {
             c_env.push(CString::new(format!("{}={}", key, value))?);
         }
-        execve(&path, &c_args, &c_env).map_err(|error| {
-            let hint = match parse_shebang(executable) {
-                Some(shebang) => format!("Does \"{}\" exist?", shebang),
-                None => "Does your interpreter exist?".to_string(),
-            };
-            format!(
-                "execve'ing {} failed with error: {}\n{}",
-                executable.to_string_lossy(),
-                error,
-                hint,
+        Ok(match interpreter {
+            Some(interpreter) => {
+                c_args.push_front(CString::new(interpreter.clone())?);
+                (
+                    CString::new(interpreter)?,
+                    c_args.into_iter().collect(),
+                    c_env,
+                )
+            }
+            None => (c_executable.clone(), c_args.into_iter().collect(), c_env),
+        })
+    }
+
+    fn format_execve_error(
+        error: nix::Error,
+        interpreter: Option<Vec<u8>>,
+        program: &Path,
+    ) -> String {
+        let (program, interpreter) = if let Some(interpreter) = interpreter {
+            (program, String::from_utf8_lossy(&interpreter).to_string())
+        } else {
+            (
+                program,
+                parse_shebang(program).unwrap_or_else(|| "your interpreter".to_string()),
             )
-        })?;
+        };
+        let hint = format!("Does {} exist?", interpreter);
+        format!(
+            "execve'ing {} failed with error: {}\n{}",
+            program.to_string_lossy(),
+            error,
+            hint,
+        )
+    }
+
+    fn execve(
+        interpreter: Option<Vec<u8>>,
```

Comment:
Maybe this can be passed in by reference? And then only cloned in the case of errors? Not sure whether that would make things better.